### DISCUSSION TITLE
Use import maps to increase cache-ability of assets

### DISCUSSION THREAD

**Discussion by sergiodxa** [remix-run/remix#3265] [Proposals]
Right now, every import in any JS/Ts file includes the hash of the file.

Which is great for caching but if a file changes, all the files importing that hashed file need to change too, right? Because if Remix doesn't do that the file they are importing will be the old one.

So hashing and creating many chunks like Remix does is great for caching, but once one file change only the files imported by that updated file will remain unchanged.

Something like this:

- A imports B
- B imports C

If B changes C will remain unchanged, but A will need to change to use the new hash for B

With [import maps](https://github.com/WICG/import-maps), Remix could inline the import map in the HTML, and files could have imports without hashes, the map will tell the browser "when you import B, actually import B-hash.js", this way a change to B will not require a change to A.

So the idea its that `<Scripts />` would render the preload link, the script of type module and a script with an import map like this one.

```html
<script type="importmap">
{
  "imports": {
    "@remix-run/react": "/build/remix-run-react-HASH.js",
    "react-router-dom: "/build/react-router-dom-HASH.js",
    "entry.client.js": "/build/entry-client-HASH.js",
    "manifest.js": /build/manifest-HASH.js",
    "~/hooks/use-something.js": "/build/hooks-use-something-HASH.js",
    "~/components/something.js": "/build/something-HASH.js"
  }
}
</script>
```

All JS files would use the keys for the imports like this:

```js
import { useSomething } from "~/hooks/use-something.js"
// code here
```

And the import map would let the browser know exactly what file to download and parse. When the file `~/hooks/use-something.js` change, Remix would create a new hashed file, and when a user does a document request it will get a new import map script with the new hash, but the files importing that file will not need to change together with the imported file to update the hash.

This should help increase the cacheability of client-side JS files, if you change a single file only that file will need to be downloaded again by the browser, every previously cached file should remain unchanged.

**Comment by kiliman** [comment]
That would be great, but unfortunately Firefox and Safari don't support import maps. If all the files are referencing the non-hash version, that would make it difficult to fallback gracefully for browsers that don't support it.

Too bad requests for static files don't automatically include the hash as an Etag. That way you can use normal file names and the Etag changes when the content does. I'm sure you can do that with Express, but don't know about the other platforms.

https://caniuse.com/import-maps

**Reply by sergiodxa** [reply]
I know there are multiple polyfills for import maps to make them work in other browser https://github.com/WICG/import-maps#community-polyfills-and-tooling.

But yeah this is something that could wait until more browsers support it.

**Comment by jfsiii** [comment]
The latest Firefox and Safari now support import maps
